File: packages/air-ts/src/model_dump.ts
/**
 * Port of `packages/core/src/air/model_dump.py` (model_to_dict) plus the
 * exporter's `json.dumps(..., indent=2, sort_keys=True) + "\n"` framing.
 *
 * Determinism contract (from the module docstring + export_golden.py):
 *   - object keys are emitted in sorted order (handled by json.dumps sort_keys,
 *     which our `dumps` reproduces), so the manual dict sorting in the oracle is
 *     redundant for keys and we don't replicate it;
 *   - LISTS are emitted in a defined order and json.dumps does NOT reorder them,
 *     so we must sort them here exactly as the oracle does:
 *       analog            sorted by subsystem id
 *       analog[].uses     sorted (by string)
 *       analog[].probes   sorted by probe id
 *       bridges           sorted by id
 *       exports           sorted by target
 *       assertions/operations/backends/... preserve document order (NOT sorted)
 *   - components are keyed by id (a map -> sorted keys via dumps); each pins and
 *     properties sub-map is also a map -> sorted by dumps.
 *
 * All string sorts use Unicode-code-point order to match Python `sorted()`.
 */

import type { SystemIR } from "./model.js";
import { type JsonValue, dumps } from "./json.js";

/** Code-point (Python `sorted`) comparator for BMP strings. */
function byCodePoint(a: string, b: string): number {
  if (a < b) return -1;
  if (a > b) return 1;
  return 0;
}

/** Build the deterministically-ordered plain object (model_to_dict). */
export function modelToObject(ir: SystemIR): JsonValue {
  // Model collections are Maps (insertion-ordered; see model.ts). `dumps` sorts
  // Map keys exactly like object keys, so serialization output is unchanged.
  const components: Record<string, JsonValue> = {};
  for (const cid of [...ir.components.keys()].sort(byCodePoint)) {
    const comp = ir.components.get(cid);
    if (comp === undefined) continue;
    // _plain(comp) then explicitly re-set pins/properties as sorted maps. Since
    // our `dumps` sorts every object's keys, plain object copies suffice; the
    // ordering is applied at serialization time.
    const pins: Record<string, JsonValue> = {};
    for (const pinName of [...comp.pins.keys()].sort(byCodePoint)) {
      const pin = comp.pins.get(pinName);
      if (pin === undefined) continue;
      pins[pinName] = {
        name: pin.name,
        net: pin.net,
        function: pin.function,
      };
    }
    const properties: Record<string, JsonValue> = {};
    for (const propName of [...comp.properties.keys()].sort(byCodePoint)) {
      properties[propName] = comp.properties.get(propName) as string;
    }
    const componentObj: Record<string, JsonValue> = {
      id: comp.id,
      type: comp.type,
      part: comp.part,
      spice_model: comp.spice_model,
      spice_subckt: comp.spice_subckt,
      value: comp.value,
      pins,
      properties,
    };
    // OMIT-WHEN-NULL for the optional <gui> hint (issue #22): every pre-#22
    // corpus design has gui=null, and dropping the key entirely (rather
    // than emitting "gui": null) preserves byte-parity with the frozen
    // model.json fixtures. Same pattern as Test.analysis (#62).
    if (comp.gui !== null) {
      componentObj["gui"] = { x: comp.gui.x, y: comp.gui.y, rot: comp.gui.rot };
    }
    components[cid] = componentObj;
  }

  const analog: JsonValue[] = [];
  for (const sub of [...ir.analog].sort((a, b) => byCodePoint(a.id, b.id))) {
    analog.push({
      id: sub.id,
      uses: [...sub.uses].sort(byCodePoint),
      probes: [...sub.probes]
        .sort((a, b) => byCodePoint(a.id, b.id))
        .map((p) => ({ id: p.id, net: p.net, quantity: p.quantity })),
    });
  }

  const bridges: JsonValue[] = [...ir.bridges]
    .sort((a, b) => byCodePoint(a.id, b.id))
    .map((b) => ({ id: b.id, type: b.type, data: b.data as unknown as JsonValue }));

  const exports: JsonValue[] = [...ir.exports]
    .sort((a, b) => byCodePoint(a.target, b.target))
    .map((e) => ({ target: e.target, enabled: e.enabled }));

  const payload: Record<string, JsonValue> = {
    name: ir.name,
    ir_version: ir.ir_version,
    metadata: { ...ir.metadata },
    requirements: ir.requirements.map((r) => ({ ...r })),
    nets: mapValues(ir.nets),
    power_domains: mapValues(ir.power_domains),
    components,
    interfaces: mapValues(ir.interfaces),
    analog,
    firmware_projects: mapValues(ir.firmware_projects),
    firmware_bindings: mapValues(ir.firmware_bindings),
    firmware_tasks: mapValues(ir.firmware_tasks),
    bridges,
    tests: testsMapValues(ir),
    simulation_profiles: mapValues(ir.simulation_profiles),
    exports,
  };
  // OMIT-WHEN-NULL for the optional inline firmware source (issue #36): every
  // design without a <firmware><source> block has firmware_source=null, and
  // dropping the key entirely (rather than emitting "firmware_source": null)
  // preserves byte-parity with the frozen model.json fixtures -- the same
  // pattern as Component.gui (#22) and Test.analysis (#62). `dumps` sorts keys,
  // so the position in this object is irrelevant. `pins` is a string[] -> JSON
  // array, mirroring the Python tuple rendered by model_dump._plain. The truthy
  // check treats both null and undefined (an omitted optional) as "absent".
  if (ir.firmware_source) {
    payload["firmware_source"] = {
      mcu: ir.firmware_source.mcu,
      language: ir.firmware_source.language,
      entry: ir.firmware_source.entry,
      pins: [...ir.firmware_source.pins],
      source: ir.firmware_source.source,
    };
  }
  return payload;
}

/**
 * Serialize the tests Map, omitting the ``analysis`` key on any Test whose
 * ``analysis`` is null (issue #62). The Python oracle does the same in
 * ``model_dump.model_to_dict`` -- a Test that never grew an <analysis> child
 * (i.e. every design in the pre-#62 corpus) dumps without the field at all so
 * the frozen model.json bytes are unchanged. A Test that DOES carry an AC
 * analysis serializes the nested object verbatim.
 */
function testsMapValues(ir: SystemIR): JsonValue {
  const out: Record<string, JsonValue> = {};
  for (const [testId, test] of ir.tests) {
    const testObj: Record<string, JsonValue> = {
      id: test.id,
      description: test.description,
      setup: test.setup as unknown as JsonValue,
      duration: test.duration,
      assertions: test.assertions as unknown as JsonValue,
    };
    if (test.analysis !== null) {
      testObj["analysis"] = { ...test.analysis };
    }
    out[testId] = testObj;
  }
  return out;
}

/**
 * Serialize a SystemIR to the byte-exact `model.json` string (trailing newline
 * included).
 */
export function serializeModel(ir: SystemIR): string {
  return dumps(modelToObject(ir));
}

/**
 * Pass a model Map through as a JsonValue. `dumps` serializes a Map exactly
 * like an object with sorted keys, and it recurses into nested Maps (a Test's
 * `setup`, a SimulationProfile's `properties`), so no per-field conversion is
 * needed and the emitted bytes are identical to the previous Record layout.
 */
function mapValues<T>(m: Map<string, T>): JsonValue {
  return m as unknown as JsonValue;
}

File: packages/core/src/air/model_dump.py
"""Deterministic typed-model serialization for the golden corpus.

``dump-model`` freezes the parsed :class:`~air.model.SystemIR` into JSON that the
TypeScript parser port (issue #7) is diffed against byte-for-byte. Determinism is
therefore the whole point of this module:

* dict keys are sorted (``json.dumps(..., sort_keys=True)``),
* every mapping in ``SystemIR`` (nets, components, pins, ...) is emitted in sorted
  key order regardless of document order,
* every list is emitted in a defined order (by id / name where one exists),
* floats are rendered via ``repr`` so the exact IEEE-754 value round-trips — but
  the typed model here is string-bearing (values like ``"330k"`` stay strings), so
  ``repr`` only matters for the rare genuine float, keeping output stable.

The output is a plain JSON object; the caller decides indentation. The corpus
exporter writes it with ``indent=2`` and a trailing newline.
"""

from __future__ import annotations

from dataclasses import fields, is_dataclass
from typing import Any

from .model import SystemIR


def _plain(value: Any) -> Any:
    """Recursively convert a value into JSON-native, deterministically ordered data.

    * dataclasses -> dict of their fields (recursively)
    * dicts       -> dict with keys sorted at serialization time (sort_keys)
    * lists/tuples -> list (order preserved; callers pre-sort where a stable key exists)
    * floats      -> repr() string is NOT used here; json handles floats, and repr()
                     equality is preserved by Python's float repr since 3.1. We keep
                     the native float so numeric consumers can read it; determinism of
                     text comes from Python's round-trip-safe float repr.
    """
    if is_dataclass(value) and not isinstance(value, type):
        return {f.name: _plain(getattr(value, f.name)) for f in fields(value)}
    if isinstance(value, dict):
        return {str(k): _plain(v) for k, v in value.items()}
    if isinstance(value, (list, tuple)):
        return [_plain(v) for v in value]
    return value


def _sorted_map(mapping: dict[str, Any]) -> dict[str, Any]:
    return {k: _plain(mapping[k]) for k in sorted(mapping)}


def _sorted_list(items: list[Any], key: str) -> list[Any]:
    return [_plain(item) for item in sorted(items, key=lambda i: getattr(i, key, ""))]


def model_to_dict(ir: SystemIR) -> dict[str, Any]:
    """Serialize a :class:`SystemIR` to a deterministically ordered dict.

    Mappings are emitted in sorted-key order and lists in a defined order so the
    output is byte-identical across runs, platforms, and Python dict-insertion order.
    Pins and probes nested inside components/subsystems are likewise sorted.
    """
    components: dict[str, Any] = {}
    for cid in sorted(ir.components):
        comp = ir.components[cid]
        comp_dict = _plain(comp)
        # Pins is a dict keyed by pin name; enforce sorted key order explicitly.
        comp_dict["pins"] = _sorted_map(comp.pins)
        comp_dict["properties"] = _sorted_map(comp.properties)
        components[cid] = comp_dict

    analog: list[Any] = []
    for sub in sorted(ir.analog, key=lambda s: s.id):
        sub_dict = _plain(sub)
        sub_dict["uses"] = sorted(sub.uses)
        sub_dict["probes"] = _sorted_list(sub.probes, "id")
        analog.append(sub_dict)

    tests_dict = _sorted_map(ir.tests)
    # A Test whose ``analysis`` field is None (the historical default -- all
    # corpus designs) must dump WITHOUT the ``analysis`` key at all, not as
    # ``"analysis": null``. This preserves byte-parity with the pre-#62 corpus
    # (which was frozen before Test grew an analysis field) so no fixture needs
    # regenerating. A Test that DOES carry an AC analysis serializes the nested
    # dict via the dataclass fallback in ``_plain``.
    for test_dict in tests_dict.values():
        if isinstance(test_dict, dict) and test_dict.get("analysis") is None:
            test_dict.pop("analysis", None)

    return {
        "name": ir.name,
        "ir_version": ir.ir_version,
        "metadata": _plain(ir.metadata),
        "requirements": [_plain(r) for r in ir.requirements],
        "nets": _sorted_map(ir.nets),
        "power_domains": _sorted_map(ir.power_domains),
        "components": components,
        "interfaces": _sorted_map(ir.interfaces),
        "analog": analog,
        "firmware_projects": _sorted_map(ir.firmware_projects),
        "firmware_bindings": _sorted_map(ir.firmware_bindings),
        "firmware_tasks": _sorted_map(ir.firmware_tasks),
        "bridges": _sorted_list(ir.bridges, "id"),
        "tests": tests_dict,
        "simulation_profiles": _sorted_map(ir.simulation_profiles),
        "exports": _sorted_list(ir.exports, "target"),
    }
